File: web/src/lib/contracts.ts
/** Wire contracts mirrored from core/contracts.py in the implementation brief. */
export type Service = "catalog" | "cart" | "checkout" | "payments";
export type FaultKind =
  "bad_config" | "null_error" | "slow_query" | "type_error";
export type Stage =
  | "alerted"
  | "investigating"
  | "cause_found"
  | "reproduced"
  | "fixes_proposed"
  | "verifying"
  | "fix_verified"
  | "no_fix_found"
  | "pr_opened"
  | "resolved";

export interface Alert {
  alert_id: string;
  service: Service;
  signal: "error_rate" | "latency_p95";
  value: number;
  threshold: number;
  started_at: string;
}
export interface Evidence {
  evidence_id: string;
  kind: "trace" | "exception" | "deploy_diff" | "code" | "metric";
  summary: string;
  detail: string;
  source_ref?: string | null;
}
export interface RootCause {
  service: Service;
  file: string;
  line?: number | null;
  commit?: string | null;
  explanation: string;
  evidence_ids: string[];
  confidence: number;
}
export interface ReproTest {
  path: string;
  code: string;
  fails_on_current?: boolean | null;
}
export interface FixCandidate {
  candidate_id: string;
  title: string;
  rationale: string;
  patch: string;
  files_touched: string[];
  origin?: "authored" | "gemini";
  snapshot_sha?: string | null;
}
export interface VerifyResult {
  candidate_id: string;
  applied: boolean;
  repro_passed: boolean;
  suite_passed: boolean;
  tests_run: number;
  tests_failed: number;
  duration_s: number;
  log_tail: string;
}
export interface FixRisk {
  action: "auto_apply" | "needs_human" | "reject";
  confidence: number;
  source: "jev" | "fallback";
}
/** Lists may be omitted on the wire: the Python model supplies empty defaults. */
export interface IncidentUpdate {
  incident_id: string;
  stage: Stage;
  at: string;
  message: string;
  evidence?: Evidence[];
  root_cause?: RootCause | null;
  candidates?: FixCandidate[];
  results?: VerifyResult[];
  pr_url?: string | null;
}
export interface VoiceCue {
  scheduling: "INTERRUPT" | "WHEN_IDLE" | "SILENT";
  facts: string;
}
export interface StartInvestigationArgs {
  service?: Service | null;
}
export interface ApproveFixArgs {
  candidate_id?: string | null;
}
export interface ExplainArgs {
  topic: "cause" | "fix" | "evidence" | "what_was_tried";
}

File: core/contracts.py
"""The wire contract shared by the incident hub and war-room UI."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

Service = Literal["catalog", "cart", "checkout", "payments"]
FaultKind = Literal["bad_config", "null_error", "slow_query", "type_error"]
Stage = Literal[
    "alerted", "investigating", "cause_found", "reproduced", "fixes_proposed",
    "verifying", "fix_verified", "no_fix_found", "pr_opened", "resolved",
]


class Alert(BaseModel):
    alert_id: str
    service: Service
    signal: Literal["error_rate", "latency_p95"]
    value: float
    threshold: float
    started_at: datetime


class Evidence(BaseModel):
    evidence_id: str
    kind: Literal["trace", "exception", "deploy_diff", "code", "metric"]
    summary: str
    detail: str
    source_ref: str | None = None


class RootCause(BaseModel):
    service: Service
    file: str
    line: int | None = None
    commit: str | None = None
    explanation: str
    evidence_ids: list[str]
    confidence: float = Field(ge=0, le=1)


class ReproTest(BaseModel):
    path: str
    code: str
    fails_on_current: bool | None = None


class FixCandidate(BaseModel):
    candidate_id: str
    title: str
    rationale: str
    patch: str
    files_touched: list[str]
    origin: Literal["authored", "gemini"] = "authored"
    snapshot_sha: str | None = None


class VerifyResult(BaseModel):
    candidate_id: str
    applied: bool
    repro_passed: bool
    suite_passed: bool
    tests_run: int
    tests_failed: int
    duration_s: float
    log_tail: str


class FixRisk(BaseModel):
    action: Literal["auto_apply", "needs_human", "reject"]
    confidence: float = Field(ge=0, le=1)
    source: Literal["jev", "fallback"]


class IncidentUpdate(BaseModel):
    incident_id: str
    stage: Stage
    at: datetime
    message: str
    evidence: list[Evidence] = Field(default_factory=list)
    root_cause: RootCause | None = None
    candidates: list[FixCandidate] = Field(default_factory=list)
    results: list[VerifyResult] = Field(default_factory=list)
    pr_url: str | None = None


class VoiceCue(BaseModel):
    scheduling: Literal["INTERRUPT", "WHEN_IDLE", "SILENT"]
    facts: str


class StartInvestigationArgs(BaseModel):
    model_config = ConfigDict(extra="forbid")
    service: Service | None = None


class ApproveFixArgs(BaseModel):
    model_config = ConfigDict(extra="forbid")
    candidate_id: str | None = None


class ExplainArgs(BaseModel):
    model_config = ConfigDict(extra="forbid")
    topic: Literal["cause", "fix", "evidence", "what_was_tried"]
